Workaround for race between wrap/proxy store

collectedmind opened the pull request:
If the wrapStore in the background takes too long to initialize it can cause the promise to never get picked up on the Store side.  This happened primarily on launch of chrome but could happen in other edge cases (reload and fast open)

This adds a dispatch on the wrapStore side when it gets initialized.  On the Store side it adds a specific listener that resolves the promise on a late initialization.

Issue: https://github.com/tshaddix/react-chrome-redux/issues/106

vhmth commented on src/wrap-store/wrapStore.js:
```diff
@@ -130,4 +130,18 @@ export default (store, {
+  /**
+   * Safety message to tabs for content scripts
+   */
+  chrome.tabs.query({}, tabs => {
```
I would rather not use the tabs API since that forces users of this lib to use that API. Unfortunately, the tabs API is also the one that leads to that big scary "this extension can do anything on all pages!" warning if you set `<all_urls>`.

collectedmind replied:
Totally get that sentiment.  Try to avoid that permission as well.  Fortunately, you only need that permission if you want to access details from the tab.  That being said open to any approach that makes the most sense with the least friction.

> You can use most chrome.tabs methods and events without declaring any permissions in the extension's manifest file. However, if you require access to the url, title, or favIconUrl properties of tabs.Tab, you must declare the "tabs" permission in the manifest, as shown below:
> From: https://developer.chrome.com/extensions/tabs

vhmth replied:
Ah got it!

vhmth commented on src/store/Store.js:
```diff
@@ -27,6 +27,7 @@ class Store {
 
     this.extensionId = extensionId; // keep the extensionId as an instance variable
     this.port = chrome.runtime.connect(this.extensionId, {name: portName});
+    this.safetyMessage = chrome.runtime.onMessage.addListener(this.safetyHandler.bind(this));
```
Instead of the safetyMessage approach, maybe we can just leverage [chrome.runtime.getBackgroundPage](https://developer.chrome.com/apps/runtime#method-getBackgroundPage). Once that resolves, we can then continue to do the port setup stuff. Thoughts? Seems like that might be simpler

collectedmind replied:
Cool.  I will play around with this idea today.  Avoiding additional messaging would be a boon for sure.

vhmth replied:
🙌

collectedmind replied:
The main blocker I see with the `chrome.runtime.getBackgroundPage` approach is that it does not support content scripts.  So the store would fail in that instance.  I am gonna look around the API a bit more and see if there is anything else that might work.

vhmth replied:
@collectedmind by "does not support content scripts", do you mean you can't call that from the content script?

collectedmind replied:
Sorry for the long turn around time, the weekend invaded.  To answer your question, yes.  You are not able to call that function from a content script.
  
![screen shot 2017-11-06 at 4 48 59 pm](https://user-images.githubusercontent.com/16119672/32471492-6e113532-c312-11e7-843f-b412b3b103a0.png)

rvolgers replied:
Regarding getBackgroundPage, the remark at the bottom of #115 might be relevant

corytheboyd commented:
This has become an issue for us as well, and we were thinking of the exact same type of solution. How do you think this is looking @vhmth?

vhmth commented:
@collectedmind https://www.useloom.com/share/1b152ead46ce4d9daf18dc0e94ce012b

🚢

vhmth commented:
@tshaddix what are your thoughts here? I can hook Loom up with this branch to see if we can still run the extension no problem with these changes.

vhmth commented:
@corytheboyd thank you for resurfacing - gotta get better about not falling off on updates here. :-P

tshaddix commented:
@corytheboyd @vhmth Hey guys just caught up on all of this. Wanted to read through the initial issue and the subsequent tests to make sure I had a full view of what was going on. Thanks for surfacing this. Super interesting. I think I've encountered it myself in the past.

As of right now, the given suggestion here makes the most sense.

@rvolgers is suggesting removing support for all content scripts which is a larger discussion that I would love to continue in #115. 

@vhmth I'm comfortable with this if you are.

tshaddix commented:
Also - the tests will need updated to include a stub for this as well as tests for the new behavior.

vhmth commented:
Once tests are passing, I'm game to test this on the Loom extension locally to see if it works!

collectedmind commented:
Sorry for the, entirely too long, absence of communication and attention on this.  I am not certain if these issues have been addressed elsewhere, or attention has been moved to a 2.0 version which by design does not need this.

If the above is not the case, I have updated the tests to support this change and added some test cases for the functionality.  Let me know if these feel adequate or if you would like some additional tests on it.

tshaddix commented:
@collectedmind No worries - we all get busy! Thank you for all the work here. From my knowledge, this is not something we've actually fixed in 2.0.0. I'd be interested in resolving the conflicts in `Store.js` and `wrapStore.test.js` and merging into the ongoing alpha if that's something you can help with.

tshaddix commented:
Note to self - don't use GitHub's inline editor for handling conflicts :D 

I will fix.

tshaddix commented:
Moving to #181
